apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -86,41 +241,92 @@ def render(element_html: str, data: pl.QuestionData) -> str:
-        return
+    # Get submitted answer or return format error if it does not exist
+    files = data["submitted_answers"].get(answer_name)
 
     # We will store the files in the submitted_answer["_files"] key,
-    # so delete the original submitted answer format to avoid
-    # duplication
+    # so delete the original submitted answer format to avoid duplication
     del data["submitted_answers"][answer_name]
 
     try:
-        parsed_files = json.loads(files)
+        parsed_files = json.loads(files)  # pyright: ignore [reportArgumentType]
     except Exception:
-        add_format_error(data, "Could not parse submitted files.")
+        add_format_error(answer_name, data, "Could not parse submitted files.")
         parsed_files = []
 
-    for x in parsed_files:
-        # Filter out any files that were not listed in file_names
-        if x.get("name", "") in required_file_names:
-            pl.add_submitted_file(data, x.get("name", ""), x.get("contents", ""))
-
-    # Validate that all required files are present
     if parsed_files is not None:
-        submitted_file_names = [x.get("name", "") for x in parsed_files]
-        missing_files = [
-            x for x in required_file_names if x not in submitted_file_names
-        ]
+        # Files with HTML characters "<" or ">" should have been rejected client-side already, so we silently filter them here
+        parsed_file_names = [f.get("name", "") for f in parsed_files]
+
+        # Split optional names into two separate lists: wildcard patterns and plain names
+        files_regex = [glob_to_regex(f) for f in file_patterns]
+        opt_files_regex = [glob_to_regex(f) for f in opt_file_patterns]
+
+        # Match submitted and accepted files 1:1 in this order: required_files, then required patterns, then optional files.
+        # All remaining files are then matched 1:n with optional patterns, and eventually discarded in there is no match
+        required_files = [x for x in parsed_file_names if x in file_names]
+        remaining_files = [x for x in parsed_file_names if x not in required_files]
+
+        pattern_files, missing_regex = match_regex_with_files(
+            files_regex,
+            remaining_files,
+            limit_1=True,
+        )
+        remaining_files = [x for x in remaining_files if x not in pattern_files]
+
+        # Validate that all required files are present
+        missing_files = [x for x in file_names if x not in parsed_file_names]
 
         if len(missing_files) > 0:
             add_format_error(
+                answer_name,
                 data,
-                "The following required files were missing: "
-                + ", ".join(missing_files),
+                f"The following required files were missing: {', '.join(missing_files)}",
             )
+
+        if len(missing_regex) > 0:
+            # Need some awkward mapping to patterns here to ensure we don't remove duplicates
+            missing_patterns = [
```

Comment:
```suggestion
            missing_patterns_string = ', '.join(
```
No need to make an intermediate list and join it into a string, you can pass this expression directly to the call to join.